### PR TITLE
AVRO-3041 : use default converters for unions/nullable properties

### PR DESCRIPTION
**Pull request by Gonyoda**
See also: https://github.com/apache/avro/pull/1088

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/AVRO-3041) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AVRO-3041

### Tests

- [x] My PR adds the following unit tests
- SerializeNullableWithNullableConverter
- SerializeNullableDefaultWithClass
- SerializeNullableDefaultWithDelegate
- SerializeNullableDefaultWithNullableDelegate
- SerializeNullableDefaultWithDoubleNullableDelegate

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.

### CODE REVIEW

File: lang/csharp/src/apache/main/Reflect/ClassCache.cs [hunk 1]
```diff
@@ -29,6 +31,8 @@ public class ClassCache
     {
         private static ConcurrentBag<IAvroFieldConverter> _defaultConverters = new ConcurrentBag<IAvroFieldConverter>();
 
+        private static ConcurrentDictionary<Tuple<Type, Type, bool>, IAvroFieldConverter> _typeDefaultConverters = new ConcurrentDictionary<Tuple<Type, Type, bool>, IAvroFieldConverter>();
```
Comment (unaware1224) [inline lang/csharp/src/apache/main/Reflect/ClassCache.cs:34]: This key is MEAN ... soo freaking many types

File: lang/csharp/src/apache/main/Reflect/ClassCache.cs [hunk 2]
```diff
@@ -48,13 +52,30 @@ private void AddClassNameMapItem(RecordSchema schema, Type dotnetClass)
...
+        }
+
         /// <summary>
         /// Add a default field converter
         /// </summary>
         /// <param name="converter"></param>
         public static void AddDefaultConverter(IAvroFieldConverter converter)
         {
             _defaultConverters.Add(converter);
+            _typeDefaultConverters.Clear();
```
Comment (unaware1224) [inline lang/csharp/src/apache/main/Reflect/ClassCache.cs:78]: Why?

File: lang/csharp/src/apache/main/Reflect/ClassCache.cs [hunk 3]
```diff
@@ -67,43 +88,34 @@ public static void AddDefaultConverter(IAvroFieldConverter converter)
...
         /// <summary>
-        /// Find a default converter
+        /// Returns the Type for a given Schema.Type
         /// </summary>
         /// <param name="tag"></param>
-        /// <param name="propType"></param>
-        /// <returns>The first matching converter - null if there isnt one</returns>
-        public IAvroFieldConverter GetDefaultConverter(Avro.Schema.Type tag, Type propType)
+        /// <returns>The type</returns>
+        private Type GetAvroTypeForTag(Avro.Schema.Type tag)
```
Comment (unaware1224) [inline lang/csharp/src/apache/main/Reflect/ClassCache.cs:99]: Thank you for splitting this up

File: lang/csharp/src/apache/main/Reflect/README.md [hunk 4]
```diff
@@ -106,6 +106,40 @@ _Example TypedFieldConverter_:
     }
 ```
 
+### Nullable primitive properties
+
+Converters for primitive types (int, long, decimal, etc) automatically work with simple nullable unions/properties with the converters defined for the 
+non-nullable type.  You can create one converter and it will work for both nullable and non-nullable properties.
+
+```csharp
+        public class StringToDecimalConverter : TypedFieldConverter<string,decimal>
```
Comment (unaware1224) [inline lang/csharp/src/apache/main/Reflect/README.md:115]: Awesome add - I wish we could bubble this to the top of the PR

File: lang/csharp/src/apache/test/Reflect/TestLogMessage.cs [hunk 5]
```diff
@@ -1,113 +0,0 @@
-/**
```
Comment (unaware1224) [inline lang/csharp/src/apache/test/Reflect/TestLogMessage.cs:1]: Why did this get deleted?

File: lang/csharp/src/apache/main/Reflect/README.md (hunk 4 above)
Comment (Gonyoda) [reply lang/csharp/src/apache/main/Reflect/README.md:115]: I wasn't sure what their policy was for including helper classes.  I think _some_ people use Fixed for decimals (a byte array) instead of string.  I prefer string.  Others might prefer Fixed.  So I wasn't sure if the library should have an opinion on it.  But I would game for adding helper converters like this - AND the Fixed so that there isn't an opinion, just a sampling of common converters.

File: lang/csharp/src/apache/test/Reflect/TestLogMessage.cs (hunk 5 above)
Comment (Gonyoda) [reply lang/csharp/src/apache/test/Reflect/TestLogMessage.cs:1]: it was renamed to TestConverters - since it's doing more than testing the LogMessage model.

File: lang/csharp/src/apache/main/Reflect/ClassCache.cs (hunk 2 above)
Comment (Gonyoda) [reply lang/csharp/src/apache/main/Reflect/ClassCache.cs:78]: I'll add a comment, but when a new converter is added, it should flush the cached default converters and check if the new converter is valid for some previous case where no converter was found.  This feels like an edge case - but it might be something that happens during some client code's unit tests.  IRL, I would expect that all the default converters are added before any serialization occurs.  But, in a testing world?  I'm betting it won't be the case.

File: lang/csharp/src/apache/main/Reflect/ClassCache.cs (hunk 1 above)
Comment (Gonyoda) [reply lang/csharp/src/apache/main/Reflect/ClassCache.cs:34]: I chose Tuple here simply so I didn't have to write my own IEquatable/HashCode code.  Tuple takes care of that for me, and Type, bool both have their own IEquatable, so it works.

**Review by unaware1224 — APPROVED** [review]
